sqlmesh/core/context.py

```diff
@@ -367,18 +371,38 @@ def __init__(
         self.environment_ttl = self.config.environment_ttl
         self.pinned_environments = Environment.sanitize_names(self.config.pinned_environments)
         self.auto_categorize_changes = self.config.plan.auto_categorize_changes
+        self.default_gateway = gateway or self.config.default_gateway_name
 
         self._connection_config = self.config.get_connection(self.gateway)
         self.concurrent_tasks = concurrent_tasks or self._connection_config.concurrent_tasks
-        self._engine_adapter = engine_adapter or self._connection_config.create_engine_adapter()
 
-        self.console = console or get_console(dialect=self._engine_adapter.dialect)
+        self._engine_adapters: t.Dict[str, EngineAdapter] = {
+            self.default_gateway: self._connection_config.create_engine_adapter()
+        }
+        self._min_concurrent_tasks = self.concurrent_tasks
+        self._snapshot_evaluator: t.Optional[SnapshotEvaluator] = None
 
-        self._test_connection_config = self.config.get_test_connection(
-            self.gateway, self.default_catalog, default_catalog_dialect=self.engine_adapter.DIALECT
-        )
+        self.console = console or get_console(dialect=self.engine_adapter.dialect)
+        self._test_connection_configs: t.Dict[str, ConnectionConfig] = {
+            self.default_gateway: self.config.get_test_connection(
+                self.gateway,
+                self.default_catalog,
+                default_catalog_dialect=self.engine_adapter.DIALECT,
+            )
+        }
 
-        self._snapshot_evaluator: t.Optional[SnapshotEvaluator] = None
+        if not gateway:
```

Comment:
my thinking was that we use the gateway to "control" if we want the project to be using the specific engine only, so when it is provided to only have this engine adapter. also since some of our integration tests run in parallel across different engines, and we use this gateway to control which engine to use, without it, these tests hang

Reply:
> my thinking was that we use the gateway to "control" if we want the project to be using the specific engine only, so when it is provided to only have this engine adapter.

Hmm, I'm thinking it's more like "choose this over the default gateway", not as much "I want **only** this adapter".

For example, you could configure a multi-gateway project in a way such that when switching your gateway to a non-default one, you'd create your db objects in a different, "staging" schema or catalog (if you made either a function of the current gateway, for example). Does this make sense?

Reply:
i don't understand what that means

Reply:
@tobymao you mean my comment? I mean if you had a model like this in a multi-gateway project:

```sql
MODEL (
  name @IF(@gateway = ..., X, Y),
  ...
); ...
```

and you wanted to do things like run `sqlmesh --gateway foo plan`, you'd still want your project to work with multiple gateways. Currently, if the `--gateway` option is passed, we won't initialize the other connections / engine adapters in the constructor, and so this use case wouldn't be supported.